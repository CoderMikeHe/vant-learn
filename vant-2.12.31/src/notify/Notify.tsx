// Utils
import { createNamespace } from '../utils';
import { inherit } from '../utils/functional';

// Mixins
import { popupMixinProps } from '../mixins/popup';

// Components
import Popup from '../popup';

// Types
import { CreateElement, RenderContext } from 'vue/types';
import { DefaultSlots } from '../utils/types';
import { PopupMixinProps } from '../mixins/popup/type';

export type NotifyProps = PopupMixinProps & {
  type: 'primary' | 'success' | 'danger' | 'warning';
  color: string;
  message: number | string;
  duration: number | string;
  className?: any;
  background: string;
};

const [createComponent, bem] = createNamespace('notify');

function Notify(
  h: CreateElement,
  props: NotifyProps,
  slots: DefaultSlots,
  ctx: RenderContext<NotifyProps>
) {
  const style = {
    color: props.color,
    background: props.background,
  };

  return (
    <Popup
      value={props.value}
      style={style}
      position="top"
      overlay={false}
      duration={0.2}
      lockScroll={false}
      class={[bem([props.type]), props.className]}
      {...inherit(ctx, true)}
    >
      {slots.default?.() || props.message}
    </Popup>
  );
}

Notify.props = {
  ...popupMixinProps,
  color: String,
  message: [Number, String],
  duration: [Number, String],
  className: null as any,
  background: String,
  getContainer: [String, Function],
  type: {
    type: String,
    default: 'danger',
  },
};

export default createComponent<NotifyProps>(Notify);
