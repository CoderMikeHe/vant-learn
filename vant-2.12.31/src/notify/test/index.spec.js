import Notify from '..';
import { later } from '../../../test';

test('create a notify', async () => {
  // should not cause error when call clear before show notify
  Notify.clear();

  const notify = Notify('test');

  await later();
  expect(notify.$el.outerHTML).toMatchSnapshot();
});

test('type prop', async () => {
  const notify = Notify({
    message: 'test',
    type: 'primary',
  });

  await later();
  expect(notify.$el.outerHTML).toMatchSnapshot();
});

test('notify disappear', async () => {
  const onClose = jest.fn();
  const notify = Notify({
    message: 'test',
    color: 'red',
    background: 'blue',
    duration: 10,
    onClose,
  });

  await later();
  expect(notify.$el.outerHTML).toMatchSnapshot();

  await later(20);
  expect(notify.$el.outerHTML).toMatchSnapshot();
  expect(onClose).toHaveBeenCalledTimes(1);

  Notify({
    message: 'text2',
    duration: 0,
  });

  await later();
  expect(notify.$el.outerHTML).toMatchSnapshot();

  Notify.clear();
  await later();
  expect(notify.$el.outerHTML).toMatchSnapshot();
});

test('set default options', () => {
  Notify.setDefaultOptions({ duration: 1000 });
  expect(Notify().duration).toEqual(1000);
  Notify.resetDefaultOptions();
  expect(Notify().duration).toEqual(3000);
  Notify.clear();
});

test('onClick prop', async () => {
  const onClick = jest.fn();
  const notify = Notify({
    message: 'test',
    onClick,
  });

  notify.$el.click();
  expect(onClick).toHaveBeenCalled();
});
