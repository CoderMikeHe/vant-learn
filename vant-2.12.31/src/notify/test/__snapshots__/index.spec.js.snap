// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create a notify 1`] = `<div class="van-popup van-popup--top van-notify van-notify--danger" style="transition-duration: 0.2s; z-index: 2001;" name="van-popup-slide-top">test</div>`;

exports[`notify disappear 1`] = `<div class="van-popup van-popup--top van-notify van-notify--danger" style="transition-duration: 0.2s; z-index: 2001; color: red; background: blue;" name="van-popup-slide-top">test</div>`;

exports[`notify disappear 2`] = `<div class="van-popup van-popup--top van-notify van-notify--danger" style="transition-duration: 0.2s; z-index: 2001; color: red; background: blue; display: none;" name="van-popup-slide-top">test</div>`;

exports[`notify disappear 3`] = `<div class="van-popup van-popup--top van-notify van-notify--danger" style="transition-duration: 0.2s; z-index: 2002;" name="van-popup-slide-top">text2</div>`;

exports[`notify disappear 4`] = `<div class="van-popup van-popup--top van-notify van-notify--danger" style="transition-duration: 0.2s; z-index: 2002; display: none;" name="van-popup-slide-top">text2</div>`;

exports[`type prop 1`] = `<div class="van-popup van-popup--top van-notify van-notify--primary" style="transition-duration: 0.2s; z-index: 2001;" name="van-popup-slide-top">test</div>`;
